src/ImageStats/Histogram.cc

```diff
@@ -6,75 +6,89 @@
+    _max_val = h.GetMaxVal();
+    _histogram.histogram_bins = h.GetHistogramBins();
 }
 
-void Histogram::join(Histogram& h) { // NOLINT
-    auto num_bins = h._hist.size();
+bool Histogram::join(Histogram& h) { // NOLINT
+    if (!ConsistencyCheck(*this, h)) {
+        fmt::print("(debug) Consistency check failed to join histograms - won't join them\n");
+        return false;
+    }
+    const int num_bins = h.GetHistogramBins().size();
+    const auto& other_bins = h.GetHistogramBins();
     ThreadManager::ApplyThreadLimit();
 #pragma omp parallel for
     for (int i = 0; i < num_bins; i++) {
-        h._hist[i] += _hist[i];
+        _histogram.histogram_bins[i] += other_bins[i];
     }
+    return true;
 }
 
-void Histogram::setup_bins() {
+void Histogram::Fill(const std::vector<float>& data) {
     std::vector<int64_t> temp_bins;
-    auto num_elements = _data.size();
+    auto num_elements = data.size();
     ThreadManager::ApplyThreadLimit();
 #pragma omp parallel
     {
         auto num_threads = omp_get_num_threads();
         auto thread_index = omp_get_thread_num();
 #pragma omp single
-        { temp_bins.resize(_num_bins * num_threads); }
+        { temp_bins.resize(_histogram.num_bins * num_threads); }
 #pragma omp for
         for (int64_t i = 0; i < num_elements; i++) {
-            auto val = _data[i];
-            if (std::isfinite(val)) {
-                int bin_number = std::clamp((int)((val - _min_val) / _bin_width), 0, _num_bins - 1);
-                temp_bins[thread_index * _num_bins + bin_number]++;
+            auto val = data[i];
+            if (_min_val <= val && val <= _max_val) {
+                int bin_number = std::clamp((int)((val - _min_val) / _histogram.bin_width), 0, _histogram.num_bins - 1);
+                temp_bins[thread_index * _histogram.num_bins + bin_number]++;
             }
         }
 #pragma omp for
-        for (int64_t i = 0; i < _num_bins; i++) {
+        for (int64_t i = 0; i < _histogram.num_bins; i++) {
             for (int t = 0; t < num_threads; t++) {
-                _hist[i] += temp_bins[_num_bins * t + i];
+                _histogram.histogram_bins[i] += temp_bins[_histogram.num_bins * t + i];
             }
         }
     }
 }
 
-HistogramResults Histogram::GetHistogram() const {
-    HistogramResults results;
-    results.num_bins = _num_bins;
-    results.bin_width = _bin_width;
-    results.bin_center = _min_val + (_bin_width / 2.0);
-    results.histogram_bins = _hist;
-    return results;
-}
+bool Histogram::ConsistencyCheck(const Histogram& a, const Histogram& b) {
+    if (a.GetNbins() != b.GetNbins()) {
+        std::printf("(debug) Histograms don't have the same number of bins: %d and %d\n", a.GetNbins(), b.GetNbins());
+        return false;
+    }
+    if (fabs(a.GetMinVal() - b.GetMinVal()) > std::numeric_limits<float>::epsilon()) {
+        std::printf("(debug) Lower histograms limits are not equal: %f and %f\n", a.GetMinVal(), b.GetMinVal());
```

Comment:
please use the new logging approach (`spdlog::debug(...)`) now that #713 has been merged

Reply:
this seems like something that should result in a `warn` level